src/artm/regularizer/decorrelator_phi.cc

```diff
@@ -37,18 +46,36 @@ bool DecorrelatorPhi::RegularizePhi(const ::artm::core::PhiMatrix& p_wt,
         core::is_member(p_wt.token(token_id).class_id, config_.class_id())) {
       // count sum of weights
       float weights_sum = 0.0f;
-      for (int topic_id = 0; topic_id < topic_size; ++topic_id) {
-        if (topics_to_regularize[topic_id])
-          weights_sum += p_wt.get(token_id, topic_id);
+      if (topic_pairs_.size() == 0) {
+        for (auto& pair : topics_to_regularize)
+          weights_sum += p_wt.get(token_id, pair.second);
       }
 
       // form the value
-      for (int topic_id = 0; topic_id < topic_size; ++topic_id) {
-        if (topics_to_regularize[topic_id]) {
-          float weight = p_wt.get(token_id, topic_id);
-          float value = static_cast<float>(- weight * (weights_sum - weight));
-          result->set(token_id, topic_id, value);
+      for (auto& pair : topics_to_regularize) {
+        if (topic_pairs_.size()) {
+          // NOTE: we regularize only topics from topics_to_regularize, and in
+          // this case they will be regularized using their topics from topic_map_,
+          // even if these topics are not presented in topics_to_regularize.
```

Comment:
To me this behavior is conter-intuitive. I would suggest to require an empty topic_name field in the config if user wants to use fisrt/second_topic_name nad value. Empty topic_name will be populated with all topics, and then your code works very straightforward.